backend/wayland/wl_seat.c

```diff
@@ -8,34 +8,188 @@
 #include <wlr/interfaces/wlr_input_device.h>
 #include <wlr/interfaces/wlr_pointer.h>
 #include <wlr/interfaces/wlr_keyboard.h>
+#include <wlr/interfaces/wlr_touch.h>
 #include <wlr/util/log.h>
 #include "backend/wayland.h"
 
-static void wlr_wl_device_destroy(struct wlr_input_device_state *state) {
+static void pointer_handle_enter(void *data, struct wl_pointer *wl_pointer,
+		uint32_t serial, struct wl_surface *surface, wl_fixed_t surface_x,
+		wl_fixed_t surface_y) {
+	struct wlr_input_device *dev = data;
+	assert(dev && dev->pointer && dev->pointer->state);
+	struct wlr_pointer_state *state = dev->pointer->state;
+	state->surface_x = wl_fixed_to_double(surface_x);
+	state->surface_y = wl_fixed_to_double(surface_y);
+}
+
+static void pointer_handle_leave(void *data, struct wl_pointer *wl_pointer,
+		uint32_t serial, struct wl_surface *surface) {
+
+}
+
+static void pointer_handle_motion(void *data, struct wl_pointer *wl_pointer,
+		uint32_t time, wl_fixed_t surface_x, wl_fixed_t surface_y) {
+	struct wlr_input_device *dev = data;
+	assert(dev && dev->pointer && dev->pointer->state);
+	struct wlr_pointer_state *state = dev->pointer->state;
+
+	double x = wl_fixed_to_double(surface_x);
+	double y = wl_fixed_to_double(surface_y);
+
+	if (x == state->surface_x && y == state->surface_y)
+		return;
+
+	struct wlr_event_pointer_motion wlr_event;
+	wlr_event.time_sec = time / 1000;
+	wlr_event.time_usec = time * 1000;
+	wlr_event.delta_x = x - state->surface_x;
+	wlr_event.delta_y = y - state->surface_y;
+	wl_signal_emit(&dev->pointer->events.motion, &wlr_event);
+
+	state->surface_x = x;
+	state->surface_y = y;
+}
+
+static void pointer_handle_button(void *data, struct wl_pointer *wl_pointer,
+		uint32_t serial, uint32_t time, uint32_t button, uint32_t state) {
+	struct wlr_input_device *dev = data;
+	assert(dev && dev->pointer);
+
+	struct wlr_event_pointer_button wlr_event;
+	wlr_event.button = button;
+	wlr_event.state = state;
+	wlr_event.time_sec = time / 1000;
+	wlr_event.time_usec = time * 1000;
+	wl_signal_emit(&dev->pointer->events.button, &wlr_event);
+}
+
+static void pointer_handle_axis(void *data, struct wl_pointer *wl_pointer,
+		uint32_t time, uint32_t axis, wl_fixed_t value) {
+	struct wlr_input_device *dev = data;
+	assert(dev && dev->pointer);
+
+	struct wlr_event_pointer_axis wlr_event;
+	wlr_event.delta = value;
+	wlr_event.orientation = axis;
+	wlr_event.time_sec = time / 1000;
+	wlr_event.time_usec = time * 1000;
+	wlr_event.source = WLR_AXIS_SOURCE_CONTINUOUS; // TODO
```

Comment:
You could just store the source information in wlr_input_device->pointer->state?